### PR TITLE
feat: kl tensorization

### PR DESCRIPTION
**Pull request by LorenzoLuccioli**
Add lemmas about the tensorization property for kl.

### CODE REVIEW

File: TestingLowerBounds/KullbackLeibler.lean [hunk 1]
```diff
@@ -453,5 +453,122 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+--if the general case turns out to be very hard to write and also to use, consider making a corollary where all the measures are probability measures and all the spaces are countabily generated
+
+--TODO: look into the implementation of product of kernels and measure spaces in the RD_it branch of mathlib, there is a structure for the product of measure spaces and some API that may be useful to generalize the chain rule
+
+--TODO: measurability should be able to solve something like this, maybe I should write this on zulip, see the file test_measurability
+example {ι : Type*} [Fintype ι] {β : ι → Type*} [∀ i, MeasurableSpace (β i)]
+    (s : (i : ι) → Set (β i)) (h : ∀ i, MeasurableSet (s i)) :
+    MeasurableSet (Set.pi Set.univ s) := by
+  -- measurability
+  exact MeasurableSet.univ_pi h
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]: Can you remove those TODO from this code? Record them somewhere else if you want to keep them (an issue for example).

For the measurability example: does it work if you tag `MeasurableSet.univ_pi` with `measurability`?

File: TestingLowerBounds/KullbackLeibler.lean [hunk 2]
```diff
@@ -453,5 +453,122 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+
+--TODO: find a place for this, and a better name
+--the hypothesis hι and hι' are not needed both, we can do with just one of them, but then the statement complains that it doesn't find the instance for the other, should we just leave it like this or find some way to circumvent it?
+--should μ be an explicit argument?
+lemma Measure.pi_map_CongrLeft {ι ι' : Type*} [hι : Fintype ι] [hι' : Fintype ι'] (e : ι ≃ ι')
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]:
```suggestion
lemma Measure.pi_map_piCongrLeft {ι ι' : Type*} [hι : Fintype ι] [hι' : Fintype ι'] (e : ι ≃ ι')
```

File: TestingLowerBounds/KullbackLeibler.lean [hunk 3]
```diff
@@ -453,5 +453,122 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+  have : e_meas ⁻¹' Set.pi Set.univ s = Set.pi Set.univ s' := by
+    ext x
+    simp only [Set.mem_preimage, Set.mem_pi, Set.mem_univ, forall_true_left, s']
+    apply (e.forall_congr _).symm
+    intro i
+    convert Iff.rfl
+    have piCongrLeft_apply_apply' :
+        (MeasurableEquiv.piCongrLeft (fun i' => β i') e) x (e i) = x i := by
+      simp only [MeasurableEquiv.piCongrLeft, MeasurableEquiv.coe_mk,
+        Equiv.piCongrLeft_apply_apply]
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]: That should be a lemma. Name: `MeasurableEquiv.piCongrLeft_apply_apply` ?

File: TestingLowerBounds/KullbackLeibler.lean [hunk 4]
```diff
@@ -453,5 +453,122 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+    convert kl_prod_two <;> tauto <;> infer_instance
+
+
+-- is it ok to state it like this or should we use a specific fintype like Fin n, so we have the cardinality defined in the statement?
+lemma kl_pi_const {ι : Type*} [hι : Fintype ι] [CountablyGenerated α] [IsProbabilityMeasure μ]
+    [IsProbabilityMeasure ν] :
+    kl (Measure.pi (fun (_ : ι) ↦ μ)) (Measure.pi (fun (_ : ι) ↦ ν)) = hι.card * kl μ ν := by
+  rw [kl_pi, Finset.sum_const, (Finset.card_eq_iff_eq_univ _).mpr rfl, EReal.nsmul_eq_mul]
+
+
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]:
```suggestion
```

File: TestingLowerBounds/KullbackLeibler.lean [hunk 5]
```diff
@@ -453,5 +453,122 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+          intro h i
+          rcases i <;> tauto
+        simp only [me.map_apply, univ_option, Finset.le_eq_subset, Finset.prod_insertNone, this,
+          Measure.prod_prod, Measure.pi_pi, mul_comm]
+      convert fDiv_map_measurableEmbedding me
+      <;> try {exact hh _} <;> infer_instance
+    rw [Fintype.sum_option, h, add_comm, ← ind_h]
+    convert kl_prod_two <;> tauto <;> infer_instance
+
+
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]:
```suggestion
```

File: TestingLowerBounds/KullbackLeibler.lean [hunk 6]
```diff
@@ -453,5 +453,122 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+  · intro ι ι' hι' e h β _ _ μ ν _ _
+    specialize h (β := fun i ↦ β (e i)) (μ := fun i ↦ μ (e i)) (ν := fun i ↦ ν (e i))
+    let hι : Fintype ι := Fintype.ofEquiv _ e.symm
+    rw [Fintype.sum_equiv e.symm _ (fun i ↦ kl (μ (e i)) (ν (e i))), ← h, kl_eq_fDiv, kl_eq_fDiv]
+    let e_meas : ((b : ι) → β (e b)) ≃ᵐ ((a : ι') → β a) :=
+      MeasurableEquiv.piCongrLeft (fun i ↦ β i) e
+    have me := MeasurableEquiv.measurableEmbedding e_meas.symm
+    convert (fDiv_map_measurableEmbedding me).symm
+    <;> try {rw [← Measure.pi_map_CongrLeft e, MeasurableEquiv.map_symm_map]}
+    <;> infer_instance
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]:
```suggestion
      <;> try {rw [← Measure.pi_map_CongrLeft e, MeasurableEquiv.map_symm_map]}
      <;> infer_instance
```

File: TestingLowerBounds/KullbackLeibler.lean [hunk 7]
```diff
@@ -453,5 +453,122 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+    intro i
+    rw [Equiv.apply_symm_apply]
+  · intro β _ _ μ ν _ _
+    rw [Measure.pi_of_empty, Measure.pi_of_empty, kl_self, Finset.univ_eq_empty, Finset.sum_empty]
+  · intro ι hι ind_h β _ _ μ ν _ _
+    specialize ind_h (β := fun i ↦ β i) (μ := fun i ↦ μ i) (ν := fun i ↦ ν i)
+    have h : kl (Measure.pi μ) (Measure.pi ν) = kl (Measure.prod (Measure.pi (fun (i : ι) ↦ μ i))
+        (μ none)) (Measure.prod (Measure.pi (fun (i : ι) ↦ ν i)) (ν none)) := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]:
```suggestion
        (μ none)) ((Measure.pi (fun (i : ι) ↦ ν i)).prod (ν none)) := by
```

File: TestingLowerBounds/KullbackLeibler.lean [hunk 8]
```diff
@@ -453,5 +453,122 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+      rw [kl_eq_fDiv, kl_eq_fDiv]
+      let e_meas : ((i : ι) → β (some i)) × β none ≃ᵐ ((i : Option ι) → β i) :=
+        MeasurableEquiv.piOptionEquivProd.{_, _, u_3} β |>.symm --TODO: when we bump mathlib remove the explicit universe level
+      have me := MeasurableEquiv.measurableEmbedding e_meas
+      have hh (ξ : (i : Option ι) → Measure (β i)) [∀ (i : Option ι), IsProbabilityMeasure (ξ i)] :
+      Measure.pi ξ = Measure.map (⇑e_meas) (Measure.prod (Measure.pi fun i ↦ ξ (some i)) (ξ none))
+          := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]: 3 remarks:
1 - indent the conclusion of the have
2 - having `:= by` alone on one line is not proper style for mathlib
3 - You can use dot notation a bit more, for example for Measure.prod
```suggestion
          Measure.pi ξ = ((Measure.pi fun i ↦ ξ (some i)).prod (ξ none)).map (⇑e_meas) := by
```
Also: do you really need the coercion arrow for `e_meas`?

File: TestingLowerBounds/KullbackLeibler.lean [hunk 9]
```diff
@@ -453,5 +453,122 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+        refine Measure.pi_eq (fun s _ ↦ ?_)
+        have : e_meas ⁻¹' Set.pi Set.univ s = (Set.pi Set.univ (fun i => s (some i))) ×ˢ (s none)
+            := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]:
```suggestion
        have : e_meas ⁻¹' Set.pi Set.univ s
            = (Set.pi Set.univ (fun i => s (some i))) ×ˢ (s none) := by
```

File: TestingLowerBounds/KullbackLeibler.lean [hunk 10]
```diff
@@ -453,5 +453,122 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+          rcases i <;> tauto
+        simp only [me.map_apply, univ_option, Finset.le_eq_subset, Finset.prod_insertNone, this,
+          Measure.prod_prod, Measure.pi_pi, mul_comm]
+      convert fDiv_map_measurableEmbedding me
+      <;> try {exact hh _} <;> infer_instance
+    rw [Fintype.sum_option, h, add_comm, ← ind_h]
+    convert kl_prod_two <;> tauto <;> infer_instance
+
+
+-- is it ok to state it like this or should we use a specific fintype like Fin n, so we have the cardinality defined in the statement?
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]: It is better like this, with a general type.

File: TestingLowerBounds/KullbackLeibler.lean (hunk 1 above)
Comment (LorenzoLuccioli) [reply TestingLowerBounds/KullbackLeibler.lean]: No, even tagging the lemma with `@[measurability]`doesn't work.

**Comment by RemyDegenne** [comment]
Please update the blueprint with entries for the new lemmas.

**Comment by RemyDegenne** [comment]
I added `kl_prod_two` and `kl_pi` to the blueprint on master. You can merge and add some `leanok` to those.

File: TestingLowerBounds/KullbackLeibler.lean [hunk 11]
```diff
@@ -453,5 +453,108 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+  · intro ι hι ind_h β _ _ μ ν _ _
+    specialize ind_h (β := fun i ↦ β i) (μ := fun i ↦ μ i) (ν := fun i ↦ ν i)
+    have h : kl (Measure.pi μ) (Measure.pi ν) = kl ((Measure.pi (fun (i : ι) ↦ μ i)).prod
+        (μ none)) ((Measure.pi (fun (i : ι) ↦ ν i)).prod (ν none)) := by
+      rw [kl_eq_fDiv, kl_eq_fDiv]
+      let e_meas : ((i : ι) → β (some i)) × β none ≃ᵐ ((i : Option ι) → β i) :=
+        MeasurableEquiv.piOptionEquivProd.{_, _, u_3} β |>.symm --TODO: when we bump mathlib remove the explicit universe level
+      have me := MeasurableEquiv.measurableEmbedding e_meas
+      have hh (ξ : (i : Option ι) → Measure (β i)) [∀ (i : Option ι), IsProbabilityMeasure (ξ i)] :
+          Measure.pi ξ = ((Measure.pi fun i ↦ ξ (some i)).prod (ξ none)).map (⇑e_meas) := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]: Turn this into a lemma? (replacing e_meas by its value). It looks like this one could be useful for similar inductions, beyond the kl case.

File: TestingLowerBounds/KullbackLeibler.lean [hunk 12]
```diff
@@ -453,5 +453,108 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+
+--TODO: put this in the right place, maybe PR to mathlib, just after MeasurableEquiv.piCongrLeft?
+lemma MeasurableEquiv.piCongrLeft_apply_apply {ι ι' : Type*} (e : ι ≃ ι') {β : ι' → Type*}
+    [∀ i', MeasurableSpace (β i')] (x : (i : ι) → β (e i)) (i : ι) :
+    (MeasurableEquiv.piCongrLeft (fun i' => β i') e) x (e i) = x i := by
+  rw [MeasurableEquiv.piCongrLeft, MeasurableEquiv.coe_mk, Equiv.piCongrLeft_apply_apply]
+
+--TODO: find a place for this, and a better name
+--the hypothesis hι and hι' are not needed both, we can do with just one of them, but then the statement complains that it doesn't find the instance for the other, should we just leave it like this or find some way to circumvent it?
+--should μ be an explicit argument?
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]: I think it's fine to keep both fintypes. Let's make `μ` explicit.
```suggestion
```

File: TestingLowerBounds/KullbackLeibler.lean [hunk 13]
```diff
@@ -453,5 +453,108 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
...
+lemma Measure.pi_map_piCongrLeft {ι ι' : Type*} [hι : Fintype ι] [hι' : Fintype ι'] (e : ι ≃ ι')
+    {β : ι' → Type*} [∀ i, MeasurableSpace (β i)] {μ : (i : ι') → Measure (β i)}
```
Comment (RemyDegenne) [inline TestingLowerBounds/KullbackLeibler.lean]:
```suggestion
    {β : ι' → Type*} [∀ i, MeasurableSpace (β i)] (μ : (i : ι') → Measure (β i))
```

**Comment by RemyDegenne** [comment]
I checked the code and blueprint, and this PR should be ready to merge after those last changes.

**Review by RemyDegenne — APPROVED** [review]

**Pull request merged**